Add CrossSite Request Forgery prevention filter

sberyozkin opened the pull request:
Fixes #8399.

This PR is based on the ideas shown by @kekbur  in the code shared at #8399 but with some modifications:
- It is a RESTEasy Reactive implementation of a `preMatching` (implicitly non-blocking) server request filter
-  `RoutingContext`'s way of checking the form fields is used and the `InputStream` is recreated
- Token verification in the filter is optional - documentation shows how to easily compare the values in the application code when dealing with the large payloads
- the filter is configurable (various cookie properties, form field name, etc).

However I've not managed to resolve 2 problems yet which might be related:
* I could not make `CsrfReactiveConfig` runtime init - the test is failing if it is runtime init saying that `CsrfReactiveConfig` has not been initialized at the time of `CsrfRequestResponseReactiveFilter`'s construction ( `CsrfReactiveConfig` is injected into the `CsrfRequestResponseReactiveFilter` constructor). I tried initializing the filter it at the runtime init time - by injecting an empty recorder which did not really help...
* I could not create an internal `non-static` `SecureRandom` field in `CsrfRequestResponseReactiveFilter` - the native test buid is failing in this case - looks like RESTEasy Reactive filters are kept in a `static` field in `RESTEasy Reactive`? So a new `SecureRandom` is created directly in the request filter method implementation right now - not a big problem as the token creation should not happen often but I wonder if it can be optimized

@geoand @mkouba Can you please review and advise on resolving these 2 issues ?

sberyozkin committed:
Add CrossSite Request Forgery prevention filter

mkouba commented:
> I could not make CsrfReactiveConfig runtime init - the test is failing if it is runtime init saying that CsrfReactiveConfig has not been initialized at the time of CsrfRequestResponseReactiveFilter's construction ( CsrfReactiveConfig is injected into the CsrfRequestResponseReactiveFilter constructor). I tried initializing the filter it at the runtime init time - by injecting an empty recorder which did not really help...

@sberyozkin @geoand Hm, when exactly is the `CsrfRequestResponseReactiveFilter` created?

sberyozkin commented:
@mkouba I'm adding it [here](https://github.com/quarkusio/quarkus/pull/27726/files#diff-3ebc92edd671b8ec5fb0ea0d904003fb9e2b029286a08fc88f38e9b154997d0cR18)

mkouba commented:
> @mkouba I'm adding it [here](https://github.com/quarkusio/quarkus/pull/27726/files#diff-3ebc92edd671b8ec5fb0ea0d904003fb9e2b029286a08fc88f38e9b154997d0cR18)

Yes, but that's not enough information - this merely adds the bean but does not say when the filter is created.

sberyozkin commented:
@mkouba I see, sure, Georgios, can you clarify please when RESTEasy Reactive does it ?

sberyozkin commented:
Yeah, I've just realized the same is the case with another reactive filter config: https://github.com/quarkusio/quarkus/blob/main/extensions/oidc-client-reactive-filter/runtime/src/main/java/io/quarkus/oidc/client/reactive/filter/runtime/OidcClientReactiveFilterConfig.java

geoand commented:
@sberyozkin have you tried using field injection?

sberyozkin commented:
Oops :-), sure, I'll try shortly

sberyozkin commented:
@geoand, @mkouba, Injecting it as a field makes no difference, here is a stack trace:
```
Caused by: java.lang.RuntimeException: Error injecting io.quarkus.csrf.reactive.runtime.CsrfReactiveConfig io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter.config
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter_Bean.create(Unknown Source)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:111)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:32)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:32)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter_Bean.get(Unknown Source)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter_Bean.get(Unknown Source)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter$GeneratedServerResponseFilter$filter_Bean.create(Unknown Source)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter$GeneratedServerResponseFilter$filter_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:111)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:32)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:32)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter$GeneratedServerResponseFilter$filter_Bean.get(Unknown Source)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter$GeneratedServerResponseFilter$filter_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:468)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:481)
	at io.quarkus.arc.impl.ArcContainerImpl$1.get(ArcContainerImpl.java:285)
	at io.quarkus.arc.impl.ArcContainerImpl$1.get(ArcContainerImpl.java:282)
	at io.quarkus.arc.runtime.BeanContainerImpl$1.create(BeanContainerImpl.java:36)
	at io.quarkus.resteasy.reactive.common.runtime.ArcBeanFactory.createInstance(ArcBeanFactory.java:27)
	at org.jboss.resteasy.reactive.server.core.startup.RuntimeInterceptorDeployment.createInterceptorInstances(RuntimeInterceptorDeployment.java:153)
	at org.jboss.resteasy.reactive.server.core.startup.RuntimeInterceptorDeployment.<init>(RuntimeInterceptorDeployment.java:65)
	at org.jboss.resteasy.reactive.server.core.startup.RuntimeDeploymentManager.deploy(RuntimeDeploymentManager.java:95)
	at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder.createDeployment(ResteasyReactiveRecorder.java:174)
	at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment1557341086.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment1557341086.deploy(Unknown Source)
	... 50 more
Caused by: javax.enterprise.inject.CreationException: Config root [io.quarkus.csrf.reactive.runtime.CsrfReactiveConfig] with config phase [RUN_TIME] not initialized yet.
	at io.quarkus.csrf.reactive.runtime.CsrfReactiveConfig_ce03c2a4e7c8cf63d1ce84a2009a32d0498e3425_Synthetic_Bean.create(Unknown Source)
	at io.quarkus.csrf.reactive.runtime.CsrfReactiveConfig_ce03c2a4e7c8cf63d1ce84a2009a32d0498e3425_Synthetic_Bean.get(Unknown Source)
	at io.quarkus.csrf.reactive.runtime.CsrfReactiveConfig_ce03c2a4e7c8cf63d1ce84a2009a32d0498e3425_Synthetic_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.CurrentInjectionPointProvider.get(CurrentInjectionPointProvider.java:60)
	... 82 more
```

It looks like the reactive filters can have a Runtime Init config at this stage which is not a PR blocker, and it should be looked it in an enhancement request, do you agree ?

What about the 2nd problem (re `SecureRandom`, see the description) ?

geoand commented:
`SecureRandom` should indeed not be a static field, this causes problems in native (which can be overcome however if necessary)

sberyozkin commented:
@geoand So if you look at the description of the 2nd problem - it does not work if I set as an instance variable.

I think I'll go ahead and open 2 issues as I believe there is nothing that can be done at this PR level to address either of the issues.

Can you please review the reactive code etc ?

geoand commented:
> Can you please review the reactive code etc ?

Sorry, I have too many bugs to look into currently, so this will have to wait.

geoand commented:
To be honest,  I don't think the implementation of `CsrfRequestResponseReactiveFilter` will work properly. What you likely need in this case is to update the handler chain with a custom `ServerRestHandler`.

sberyozkin commented:
@geoand I can ask someone else to review if you are busy, np.

>To be honest, I don't think the implementation of CsrfRequestResponseReactiveFilter will work properly. What you likely need in this case is to update the handler chain with a custom ServerRestHandler.

I have the tests passing and I do not understand your comment - the implementation seems sound to me. I'm OK with reworking it as `ServerRestHandler` - but is it really necessary, should I do if the users spot some actual issues later ?

geoand commented:
Nevermind my comment about `ServerRestHandler`, what you are doing can work, however I strongly suggest the filters be rewritten to return `Uni<Void>` so you can avoid pausing the resuming the execution of the handler chain manually.

sberyozkin commented:
@geoand I'm open to all your expert suggestions, I'll try `Uni<Void>` for sure, thanks for a hint

geoand commented:
Great, thanks!

My point for suggesting that is that a filter should just setup a `Uni` chain and never have to subscript to it and never have to deal with suspending and resuming the processing - this will automatically be done by RESTEasy Reactive.

Other than that, the implementation looks fine.

sberyozkin commented:
@geoand I've added `Uni<Void>` and it really looks nice and neat, thanks, and I think we should doc more how to do it - I can give it a try a bit later as supporting asynchronous filters is a great feature.
I've also progressed with supporting `RUN_TIME` config  - right now it is not cool, the way it is done (I need to keep a `ConfigGroup` so that `quarkus.*` namespace is recognized) and duplicate the same in the filter constructor - I can leave with it for sure as the users won't notice, the important thing the RUN_TIME config is supported. So I'm happy enough.

But I've also just seen @mkouba's suggestion - so I can give it a quick try

thanks

sberyozkin commented:
I've squashed the last commit so that if Martin's suggestion works I can keep it as a separate commit to review

sberyozkin commented:
Now I'm really happy with the way I can handle the runtime config in the filter :-), @geoand does it look OK to you ?

geoand commented:
I'll have a look tomorrow

geoand commented on docs/src/main/asciidoc/security-csrf-prevention.adoc:
```diff
@@ -0,0 +1,200 @@
+    Template csrfToken; <1>
+
+    @Inject
+    RoutingContext routingContext;
```
I would very much like to avoid showing injecting the `RoutingContext`. Users should almost never have to do this

sberyozkin replied:
@geoand Sure, I've added it to show how one can check that the filter has actually been in action to double protect if needed, but I guess just having a note there that a `RoutingContext` attribute is available should be enough - if needed a custom user filter can be added to check it...

sberyozkin replied:
I just removed this code fragment, users should indeed not be dealing with it to check this property...

sberyozkin commented on extensions/csrf-reactive/runtime/src/main/java/io/quarkus/csrf/reactive/runtime/CsrfRequestResponseReactiveFilter.java:
```diff
@@ -0,0 +1,248 @@
+                            String csrfToken = form.get(config.get().formFieldName);
+                            if (csrfToken == null) {
+                                LOG.debug("CSRF token is not found");
+                                requestContext.abortWith(Response.status(400).build());
```
@geoand have a look please if it looks better

geoand replied:
Looks good

geoand commented:
I think it would be best to squash the commits

sberyozkin commented:
Sure

sberyozkin commented:
Though I'd like to add one more config option, for the CSRF cookie optionally not to be created per every new GET request - there is an assumption there an HTML form will be returned for every one but in real applications it won't be the case, will do all tomorrow, thanks

sberyozkin commented:
I've just noticed a perf improvement suggestion from @mkouba as well, the config should be accessed via a local var, will take care of it as well

geoand reviewed (approved):
LGTM

sberyozkin commented:
I updated the code to use a local variable to access the config instance, and added 2 more config properties, one allowing to avoid creating the cookies for all the requests (this one is also checked in the test), and one allowing to pass through non-url-form-encoded POSTs as a complex application may have all sort of interactions with the outside world, and POST JSONs should not ve really blocked if a resource method exists, in the latter case the CORS filter + XHTMLRequest may be handling the CSRF risks separately...

I think it is probably all for now, it can be tricky to implement such a feature covering all the variations from the very start, but we'll have a good enough base...

sberyozkin commented:
Looks like a window build has just died, have been waiting for it for all day to finish :-)

sberyozkin commented:
I'll just go ahead and merge to give it a couple of days to settle, as I'm off tomorrow as well :-)

FroMage commented:
Wait a minute: this should also support multipart forms! https://github.com/quarkusio/quarkus/issues/28379
